### PR TITLE
[FSDP] Default `limit_all_gathers=True`

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104900

This PR defaults to `limit_all_gathers=True`.

I included a `record_function()` for the rate limiter synchronization to help with user confusion on the gap in the pre-forward:
Screenshot 2023-07-10 at 3 28 18 PM

### CODE REVIEW

**Review by fegin — APPROVED** [review]
LGTM, but should we wait for the internal diffs are landed before this PR is landed?

**Review by fduwjj — COMMENTED** [review]
Nice!!

**Comment by awgu** [comment]
Internal config diff is landing.

**Pull request closed without merging**
